From d305bd66fecf1630eea759cf7c3e7acb5b443c32 Mon Sep 17 00:00:00 2001
From: Ricardo Neri <ricardo.neri@ti.com>
Date: Sun, 26 Jun 2011 04:00:28 -0500
Subject: [PATCH 3/4] OMAP4: HDMI: Add transition states to ACR WA state machine

The HDMI driver cannot know what will be the status of the SYSM3
ACR WA until it receives the ACK. This patch adds to transition states
to reflect states in which A9 has requested to stop/start the ACR WA.
If a second request arrives while the first request is ACK'ed and then
the second request will be made.

Change-Id: I809ca7b9e48a7044fef6dc28baa5871a83ea6fd8
Signed-off-by: Ricardo Neri <ricardo.neri@ti.com>

squash

Change-Id: Ibb41402a15ae7d020962494e42f9e33e694f84a0
---
 arch/arm/plat-omap/hdmi_lib.c              |   10 ++++++++--
 arch/arm/plat-omap/include/plat/hdmi_lib.h |    2 ++
 2 files changed, 10 insertions(+), 2 deletions(-)

Index: kernel/arch/arm/plat-omap/hdmi_lib.c
===================================================================
--- kernel.orig/arch/arm/plat-omap/hdmi_lib.c	2011-07-01 18:11:16.322153393 +0530
+++ kernel/arch/arm/plat-omap/hdmi_lib.c	2011-07-01 18:11:25.979648195 +0530
@@ -1463,15 +1463,18 @@
 	}
 	return 0;
 }
+
 int hdmi_lib_start_acr_wa(void)
 {
 	int ret = 0;
 	mutex_lock(&hdmi.wa_state_lock);
-	if (hdmi.wa_state == HDMI_WA_STATE_STARTED) {
+	if (hdmi.wa_state == HDMI_WA_STATE_STARTED ||
+			hdmi.wa_state == HDMI_WA_STATE_TRANSIT_TO_START) {
 		mutex_unlock(&hdmi.wa_state_lock);
 		return ret;
 	}
 
+	hdmi.wa_state = HDMI_WA_STATE_TRANSIT_TO_START;
 	mutex_unlock(&hdmi.wa_state_lock);
 	ret = hdmi_lib_acr_wa_send_event(hdmi.cts_interval);
 	if (!ret) {
@@ -1481,15 +1484,18 @@
 	}
 	return ret;
 }
+
 int hdmi_lib_stop_acr_wa(void)
 {
 	int ret = 0;
 	mutex_lock(&hdmi.wa_state_lock);
-	if (hdmi.wa_state== HDMI_WA_STATE_STOPPED) {
+	if (hdmi.wa_state== HDMI_WA_STATE_STOPPED ||
+			hdmi.wa_state == HDMI_WA_STATE_TRANSIT_TO_STOP) {
 		mutex_unlock(&hdmi.wa_state_lock);
 		return ret;
 	}
 
+	hdmi.wa_state = HDMI_WA_STATE_TRANSIT_TO_STOP;
 	mutex_unlock(&hdmi.wa_state_lock);
 	ret = hdmi_lib_acr_wa_send_event(0);
 	if (!ret) {
Index: kernel/arch/arm/plat-omap/include/plat/hdmi_lib.h
===================================================================
--- kernel.orig/arch/arm/plat-omap/include/plat/hdmi_lib.h	2011-07-01 18:11:15.339654711 +0530
+++ kernel/arch/arm/plat-omap/include/plat/hdmi_lib.h	2011-07-01 18:11:17.212149117 +0530
@@ -478,7 +478,9 @@
 	HDMI_WA_STATE_NOTIFY_WAIT_FOR_IPC,
 	HDMI_WA_STATE_NOTIFY_EVENT_REG,
 	HDMI_WA_STATE_STOPPED,
+	HDMI_WA_STATE_TRANSIT_TO_STOP,
 	HDMI_WA_STATE_STARTED,
+	HDMI_WA_STATE_TRANSIT_TO_START,
 };
 #endif
 
